oaknational/oak-ai-lesson-assistant
# fix: chat streaming and loading states (WIP)

## Description

There's a lot in here. I suggest we break up into separate PRs once we've reviewed the overall approach.

For reference on the problems this is looking to address, see [this Notion page](https://www.notion.so/oaknationalacademy/Main-Structured-Output-10626cc4e1b180aaa99ed1ed2fdcbe5e?pvs=4).

- This addresses various problems we have had with the chat experience, prompt, data loading and lesson plan completion

UI/UX:
- Fixes some small issues with the chat forms where images were lazy-loaded (use hidden/inline-block instead of mounting/unmounting)
- Adds a new queuing scrolling manager that brings the completed items into view as soon as they are ready. The current behaviour distributes this logic amongst the DropDopSection components and it's unpredictable how they behave.
- Adjusts the behaviour of the Continue button so that we can use it to fill in any missing sections and make it clear to the LLM which sections we want to generate next
- Allows the front end to rescue from scenarios where there are missing sections. The "Continue" button now includes the text of the next section group section names or any missing sections which need to be generated. This should mean that you can always complete a lesson if it seems to get stuck or miss a section.
- Rather than "Working on it…" it displays the section that is being streamed so there is more feedback to the user

Code structure:
- Replaces the tempLessonPlan logic with a LessonPlanManager
- Improves type-safety by asserting the section keys we are using are valid keys in the lesson plan, rather than just plain strings

Streaming / persistence:
- Fixes detection of parts while content is streaming in
- Reduces how often we have to Zod parse and patch the incoming message patches down to ~once per valid message part
- Adds a "status":"complete" to each generated message part. We use this to detect when the actual content of the message part has finished streaming in.
- Adds an "iteration" property which increments each time the chat is persisted. This allows us not to overwrite the chat with a past version if for some reason we have a stale cache for a request or there is a race condition

Prompt:
- Alters the prompt based on feedback from Open AI to put your cacheable / most important instructions at the top of the prompt. This now filters out steps that have previous been completed so that you aren't repeatedly asked about basedOn etc.
- Fixes the basedOn problem (or at least makes it a lot better!) where Aila automatically picks a lesson to base the desired lesson on, but without the user choosing from the list of options
- Fixes the issue where there are sometimes long text responses
- Reorders the Aila prompt so that we have instructions about the intended task at the top and then at the end constraints about how to respond to users. This seems to fix issues around text length / inclusion of summaries etc.
- Adds in HB's recent prompt work to add definitions for missing sections
- Puts more emphasis on the Additional Materials and now generates content rather than "None"

Testing / local dev:
- Adjusts the tests to check that content comes back in section groups rather than individually
- Adds more data-test attributes so that we can check the streaming state and state of the lesson plan in tests
- Rewrites the Romans test to ensure that the chat comes back in groups and validates that the section count increases over time
- Adds a display for local dev to show you what state the chat is in and which section is streaming
- Removes some CSP warnings in local development / test

Todo

[ ] Make sure we can still load existing chats
[ ] Update Playwright tests to ensure the new flow works as expected in E2E tests
[ ] Find out the spec for when lesson plan sections should be popped down - there's some logic about mobile that I haven't quite understood with this
[ ] Find out the spec (or change it) for which sections should be displayed in groups. This seems to be buggy / incorrect logic at the moment

## Issue(s)

Fixes testing issues in [this Notion ticket](https://www.notion.so/oaknationalacademy/Main-Structured-Output-10626cc4e1b180aaa99ed1ed2fdcbe5e?pvs=4).

## How to test

1. Start a chat
2. While each section streams in, note that you get a spinner, the content appears once ready, and the scroll position updates
3. You'll see a new status display at the top left. This is only shown on local development and staging so you don't need to give feedback on it. It displays the current chat iteration which increments each time the chat is persisted, as well as the streaming status
4. Note that there is no longer a white flash where the lesson plan gets emptied and then reset
5. Note that as sections stream in you are shown each one in turn but only when it has text, rather than seeing a spinner and then text for each one
6. Scroll down while a long section is loading in and you'll see the working section with a spinner
7. Check that responses with and without choosing a basedOn work as expected
8. Check that sections are generated in groups as per the design. Eg. cycles all together rather than one at a time
9. Check that a "basedOn" is not automatically set based on the initial input
10. Check the language to make sure it is in line with the spec for chat responses being brief
